app/gui/src/presenter/graph/widgets.rs

```diff
@@ -0,0 +1,352 @@
+                let vis_metadata = Self::visualization_metadata(query);
+
+                let manager = self.manager.clone_ref();
+                trace!("Registering widget visualization for {}", req.target_expression);
+                manager.request_visualization(req.target_expression, vis_metadata);
+                None
+            }
+        }
+    }
+
+    /// Remove all widget queries of given node that are connected to expression not on this list.
+    fn retain_node_expressions(&mut self, node_id: ViewNodeId, expressions: &HashSet<ast::Id>) {
+        let registered = self.widgets_of_node.get_mut(&node_id);
+        if let Some(registered) = registered {
+            registered.retain(|expr_id| {
+                let retained = expressions.contains(expr_id);
+                if !retained {
+                    self.manager.remove_visualization(*expr_id);
+                }
+                retained
+            });
+        }
+    }
+
+    fn remove_node(&mut self, node_id: ViewNodeId) {
+        if let Some(registered) = self.widgets_of_node.remove(&node_id) {
+            registered.into_iter().for_each(|expr_id| self.manager.remove_visualization(expr_id));
+        }
+    }
+
+    fn visualization_metadata(query: &QueryData) -> Metadata {
+        let relevant_arguments = query.arguments.split_first().map_or_default(|(_self, args)| args);
+        let arguments: Vec<Code> = vec![
+            Self::escape_visualization_argument(&query.method_name).into(),
+            Self::arg_sequence(relevant_arguments).into(),
+        ];
+
+        let preprocessor = visualization::instance::PreprocessorConfiguration {
+            module:    WIDGET_VISUALIZATION_MODULE.into(),
+            method:    WIDGET_VISUALIZATION_METHOD.into(),
+            arguments: Rc::new(arguments),
+        };
+        Metadata { preprocessor }
+    }
+
+    fn escape_visualization_argument(arg: &str) -> String {
+        let segment = ast::SegmentPlain { value: arg.into() };
+        let text = ast::TextLineRaw { text: vec![segment.into()] };
+        text.repr()
+    }
+
+    fn arg_sequence(args: &[ImString]) -> String {
+        let mut buffer = String::from("[");
+        for (i, arg) in args.iter().enumerate() {
+            if i > 0 {
+                buffer.push_str(", ");
+            }
+            buffer.push_str(&Self::escape_visualization_argument(arg));
+        }
+        buffer.push(']');
+        buffer
+    }
+}
+
+/// ===============================
+/// === WidgetVisualizationData ===
+/// ===============================
+
+/// type representing the data received from the widget visualization for single widget.
+#[derive(Debug, serde::Deserialize)]
+struct WidgetVisualizationData {
+    constructor: widget::Kind,
+    display:     WidgetVisualizationDataDisplay,
+    values:      Vec<String>,
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct WidgetVisualizationDataDisplay {
+    constructor: widget::Display,
+}
```

Comment:
What this struct is for? Why do we have it and not just use `widget::Display` directly in `WidgetVisualizationData`? Any explanation in the code would be appreciated.